// n8n-nodes-base/dist/credentials/MicrosoftAzureMonitorOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicrosoftAzureMonitorOAuth2Api = void 0;
class MicrosoftAzureMonitorOAuth2Api {
    constructor() {
        this.name = 'microsoftAzureMonitorOAuth2Api';
        this.displayName = 'Microsoft Azure Monitor OAuth2 API';
        this.extends = ['microsoftOAuth2Api'];
        this.documentationUrl = 'microsoftazuremonitor';
        this.icon = 'file:icons/Microsoft.svg';
        this.httpRequestNode = {
            name: 'Microsoft Azure Monitor',
            docsUrl: 'https://learn.microsoft.com/en-us/azure/azure-monitor/logs/api/request-format',
            apiBaseUrlPlaceholder: 'https://api.loganalytics.azure.com/v1/workspaces/[workspace_id]/query',
        };
        this.properties = [
            {
                displayName: 'Tenant ID',
                required: true,
                name: 'tenantId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Resource',
                name: 'resource',
                type: 'options',
                options: [
                    {
                        name: 'Azure Log Analytics',
                        value: 'https://api.loganalytics.azure.com',
                    },
                    {
                        name: 'Log Analytics',
                        value: 'https://api.loganalytics.io',
                    },
                    {
                        name: 'Azure Monitor',
                        value: 'https://monitor.azure.com',
                    },
                    {
                        name: 'Azure Management',
                        value: 'https://management.azure.com',
                    },
                ],
                default: 'https://api.loganalytics.azure.com',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: '=https://login.microsoftonline.com/{{$self["tenantId"]}}/oauth2/authorize',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: '=https://login.microsoftonline.com/{{$self["tenantId"]}}/oauth2/token',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '=resource={{$self["resource"]}}',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: '',
            },
        ];
    }
}
exports.MicrosoftAzureMonitorOAuth2Api = MicrosoftAzureMonitorOAuth2Api;
