// n8n-core/dist/execution-engine/triggers-and-pollers.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggersAndPollers = void 0;
const di_1 = require("@n8n/di");
const n8n_workflow_1 = require("n8n-workflow");
let TriggersAndPollers = class TriggersAndPollers {
    async runTrigger(workflow, node, getTriggerFunctions, additionalData, mode, activation) {
        const triggerFunctions = getTriggerFunctions(workflow, node, additionalData, mode, activation);
        const nodeType = workflow.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
        if (!nodeType.trigger) {
            throw new n8n_workflow_1.ApplicationError('Node type does not have a trigger function defined', {
                extra: { nodeName: node.name },
                tags: { nodeType: node.type },
            });
        }
        if (mode === 'manual') {
            const triggerResponse = await nodeType.trigger.call(triggerFunctions);
            triggerResponse.manualTriggerResponse = new Promise((resolve, reject) => {
                triggerFunctions.emit = ((resolveEmit) => (data, responsePromise, donePromise) => {
                    additionalData.hooks.hookFunctions.sendResponse = [
                        async (response) => {
                            if (responsePromise) {
                                responsePromise.resolve(response);
                            }
                        },
                    ];
                    if (donePromise) {
                        additionalData.hooks.hookFunctions.workflowExecuteAfter?.unshift(async (runData) => {
                            return donePromise.resolve(runData);
                        });
                    }
                    resolveEmit(data);
                })(resolve);
                triggerFunctions.emitError = ((rejectEmit) => (error, responsePromise) => {
                    additionalData.hooks.hookFunctions.sendResponse = [
                        async () => {
                            if (responsePromise) {
                                responsePromise.reject(error);
                            }
                        },
                    ];
                    rejectEmit(error);
                })(reject);
            });
            return triggerResponse;
        }
        return await nodeType.trigger.call(triggerFunctions);
    }
    async runPoll(workflow, node, pollFunctions) {
        const nodeType = workflow.nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
        if (!nodeType.poll) {
            throw new n8n_workflow_1.ApplicationError('Node type does not have a poll function defined', {
                extra: { nodeName: node.name },
                tags: { nodeType: node.type },
            });
        }
        return await nodeType.poll.call(pollFunctions);
    }
};
exports.TriggersAndPollers = TriggersAndPollers;
exports.TriggersAndPollers = TriggersAndPollers = __decorate([
    (0, di_1.Service)()
], TriggersAndPollers);
