// n8n-editor-ui/dist/assets/TestDefinitionEditView-CwoHUF5w.js
import { hS as useTestDefinitionStore, r as ref, q as computed, d as defineComponent, c as openBlock, h as createElementBlock, F as Fragment, j as createBaseVNode, t as toDisplayString, n as normalizeClass, i as createVNode, e as createBlock, l as unref, g as useI18n, m as resolveComponent, _ as _export_sfc, f as createCommentVNode, w as withCtx, k as createTextVNode, hw as mergeModels, hx as useModel, s as renderSlot, hT as MetricsChart, hU as TestRunsTable, T as useWorkflowsStore, bo as useNodeTypesStore, U as useRoute, b as useRouter, o as onMounted, aP as N8nTooltip, C as createEventBus, aa as useCssModule, ax as withDirectives, ay as vShow, hV as ElCollapseTransition, y as nextTick, A as renderList, hK as _sfc_main$c, hW as __unplugin_components_0$2, hX as SAMPLE_EVALUATION_WORKFLOW, hY as NODE_PINNING_MODAL_KEY, hZ as __unplugin_components_2, J as useDebounce, a as useToast, hI as useAnnotationTagsStore, K as useUIStore, at as useExecutionsStore, V as VIEWS, H as watch } from "./index-DablXALM.js";
import { f as useCanvasOperations, d as useVueFlow, e as useCanvasMapping, _ as __unplugin_components_0$1 } from "./useCanvasOperations-CEPzjX8I.js";
import "./usePinnedData-BEOJK2e-.js";
function useTestDefinitionForm() {
  const evaluationsStore = useTestDefinitionStore();
  const state = ref({
    name: {
      value: `My Test ${evaluationsStore.allTestDefinitions.length + 1}`,
      tempValue: "",
      isEditing: false
    },
    tags: {
      value: [],
      tempValue: [],
      isEditing: false
    },
    description: {
      value: "",
      tempValue: "",
      isEditing: false
    },
    evaluationWorkflow: {
      mode: "list",
      value: "",
      __rl: true
    },
    metrics: [],
    mockedNodes: []
  });
  const isSaving = ref(false);
  const fields = ref({});
  const editableFields = computed(() => ({
    name: state.value.name,
    tags: state.value.tags,
    description: state.value.description
  }));
  const loadTestData = async (testId) => {
    try {
      await evaluationsStore.fetchAll({ force: true });
      const testDefinition = evaluationsStore.testDefinitionsById[testId];
      if (testDefinition) {
        const metrics = await evaluationsStore.fetchMetrics(testId);
        state.value.description = {
          value: testDefinition.description ?? "",
          isEditing: false,
          tempValue: ""
        };
        state.value.name = {
          value: testDefinition.name ?? "",
          isEditing: false,
          tempValue: ""
        };
        state.value.tags = {
          isEditing: false,
          value: testDefinition.annotationTagId ? [testDefinition.annotationTagId] : [],
          tempValue: []
        };
        state.value.evaluationWorkflow = {
          mode: "list",
          value: testDefinition.evaluationWorkflowId ?? "",
          __rl: true
        };
        state.value.metrics = metrics;
        state.value.mockedNodes = testDefinition.mockedNodes ?? [];
      }
    } catch (error) {
      console.error("Failed to load test data", error);
    }
  };
  const createTest = async (workflowId) => {
    if (isSaving.value) return;
    isSaving.value = true;
    try {
      const params = {
        name: state.value.name.value,
        workflowId,
        description: state.value.description.value
      };
      return await evaluationsStore.create(params);
    } finally {
      isSaving.value = false;
    }
  };
  const deleteMetric = async (metricId, testId) => {
    await evaluationsStore.deleteMetric({ id: metricId, testDefinitionId: testId });
    state.value.metrics = state.value.metrics.filter((metric) => metric.id !== metricId);
  };
  const updateMetrics = async (testId) => {
    const promises = state.value.metrics.map(async (metric) => {
      if (!metric.name) return;
      if (!metric.id) {
        const createdMetric = await evaluationsStore.createMetric({
          name: metric.name,
          testDefinitionId: testId
        });
        metric.id = createdMetric.id;
      } else {
        await evaluationsStore.updateMetric({
          name: metric.name,
          id: metric.id,
          testDefinitionId: testId
        });
      }
    });
    isSaving.value = true;
    await Promise.all(promises);
    isSaving.value = false;
  };
  const updateTest = async (testId) => {
    if (isSaving.value) return;
    isSaving.value = true;
    try {
      if (!testId) {
        throw new Error("Test ID is required for updating a test");
      }
      const params = {
        name: state.value.name.value,
        description: state.value.description.value
      };
      if (state.value.evaluationWorkflow.value) {
        params.evaluationWorkflowId = state.value.evaluationWorkflow.value.toString();
      }
      const annotationTagId = state.value.tags.value[0];
      if (annotationTagId) {
        params.annotationTagId = annotationTagId;
      }
      if (state.value.mockedNodes.length > 0) {
        params.mockedNodes = state.value.mockedNodes;
      }
      const response = await evaluationsStore.update({ ...params, id: testId });
      return response;
    } finally {
      isSaving.value = false;
    }
  };
  function startEditing(field) {
    const fieldObj = editableFields.value[field];
    if (fieldObj.isEditing) {
      return;
    }
    if (Array.isArray(fieldObj.value)) {
      fieldObj.tempValue = [...fieldObj.value];
    } else {
      fieldObj.tempValue = fieldObj.value;
    }
    fieldObj.isEditing = true;
  }
  function saveChanges(field) {
    const fieldObj = editableFields.value[field];
    fieldObj.value = Array.isArray(fieldObj.tempValue) ? [...fieldObj.tempValue] : fieldObj.tempValue;
    fieldObj.isEditing = false;
  }
  function cancelEditing(field) {
    const fieldObj = editableFields.value[field];
    if (Array.isArray(fieldObj.value)) {
      fieldObj.tempValue = [...fieldObj.value];
    } else {
      fieldObj.tempValue = fieldObj.value;
    }
    fieldObj.isEditing = false;
  }
  function handleKeydown(event, field) {
    if (event.key === "Escape") {
      cancelEditing(field);
    } else if (event.key === "Enter" && !event.shiftKey) {
      event.preventDefault();
      saveChanges(field);
    }
  }
  return {
    state,
    fields,
    isSaving: computed(() => isSaving.value),
    deleteMetric,
    updateMetrics,
    loadTestData,
    createTest,
    updateTest,
    startEditing,
    saveChanges,
    cancelEditing,
    handleKeydown
  };
}
const _sfc_main$b = /* @__PURE__ */ defineComponent({
  __name: "TestNameInput",
  props: {
    modelValue: {},
    startEditing: { type: Function },
    saveChanges: { type: Function },
    handleKeydown: { type: Function }
  },
  emits: ["update:modelValue"],
  setup(__props) {
    const locale = useI18n();
    return (_ctx, _cache) => {
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_N8nInput = resolveComponent("N8nInput");
      return openBlock(), createElementBlock("h2", {
        class: normalizeClass(_ctx.$style.title)
      }, [
        !_ctx.modelValue.isEditing ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
          createBaseVNode("span", {
            class: normalizeClass(_ctx.$style.titleText)
          }, toDisplayString(_ctx.modelValue.value), 3),
          createVNode(_component_n8n_icon_button, {
            class: normalizeClass(_ctx.$style.editInputButton),
            icon: "pen",
            type: "tertiary",
            onClick: _cache[0] || (_cache[0] = ($event) => _ctx.startEditing("name"))
          }, null, 8, ["class"])
        ], 64)) : (openBlock(), createBlock(_component_N8nInput, {
          key: 1,
          ref: "nameInput",
          "data-test-id": "evaluation-name-input",
          "model-value": _ctx.modelValue.tempValue,
          type: "text",
          placeholder: unref(locale).baseText("testDefinition.edit.namePlaceholder"),
          "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => _ctx.$emit("update:modelValue", { ..._ctx.modelValue, tempValue: $event })),
          onBlur: _cache[2] || (_cache[2] = () => _ctx.saveChanges("name")),
          onKeydown: _cache[3] || (_cache[3] = (e) => _ctx.handleKeydown(e, "name"))
        }, null, 8, ["model-value", "placeholder"]))
      ], 2);
    };
  }
});
const title$1 = "_title_1g2qi_123";
const titleText = "_titleText_1g2qi_133";
const editInputButton$2 = "_editInputButton_1g2qi_140";
const style0$a = {
  title: title$1,
  titleText,
  editInputButton: editInputButton$2
};
const cssModules$a = {
  "$style": style0$a
};
const TestNameInput = /* @__PURE__ */ _export_sfc(_sfc_main$b, [["__cssModules", cssModules$a]]);
const _sfc_main$a = /* @__PURE__ */ defineComponent({
  __name: "DescriptionInput",
  props: {
    modelValue: {},
    startEditing: { type: Function },
    saveChanges: { type: Function },
    handleKeydown: { type: Function }
  },
  emits: ["update:modelValue"],
  setup(__props) {
    const locale = useI18n();
    return (_ctx, _cache) => {
      const _component_n8n_icon = resolveComponent("n8n-icon");
      const _component_N8nText = resolveComponent("N8nText");
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_N8nInput = resolveComponent("N8nInput");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.description)
      }, [
        !_ctx.modelValue.isEditing ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
          createBaseVNode("span", {
            class: normalizeClass(_ctx.$style.descriptionText),
            onClick: _cache[0] || (_cache[0] = ($event) => _ctx.startEditing("description"))
          }, [
            _ctx.modelValue.value.length === 0 ? (openBlock(), createBlock(_component_n8n_icon, {
              key: 0,
              class: normalizeClass(_ctx.$style.icon),
              icon: "plus",
              color: "text-light",
              size: "medium"
            }, null, 8, ["class"])) : createCommentVNode("", true),
            createVNode(_component_N8nText, { size: "medium" }, {
              default: withCtx(() => [
                createTextVNode(toDisplayString(_ctx.modelValue.value.length > 0 ? _ctx.modelValue.value : "Add a description"), 1)
              ]),
              _: 1
            })
          ], 2),
          createVNode(_component_n8n_icon_button, {
            class: normalizeClass(_ctx.$style.editInputButton),
            icon: "pen",
            type: "tertiary",
            onClick: _cache[1] || (_cache[1] = ($event) => _ctx.startEditing("description"))
          }, null, 8, ["class"])
        ], 64)) : (openBlock(), createBlock(_component_N8nInput, {
          key: 1,
          ref: "descriptionInput",
          "data-test-id": "evaluation-description-input",
          "model-value": _ctx.modelValue.tempValue,
          type: "textarea",
          placeholder: unref(locale).baseText("testDefinition.edit.descriptionPlaceholder"),
          "onUpdate:modelValue": _cache[2] || (_cache[2] = ($event) => _ctx.$emit("update:modelValue", { ..._ctx.modelValue, tempValue: $event })),
          onBlur: _cache[3] || (_cache[3] = () => _ctx.saveChanges("description")),
          onKeydown: _cache[4] || (_cache[4] = (e) => _ctx.handleKeydown(e, "description"))
        }, null, 8, ["model-value", "placeholder"]))
      ], 2);
    };
  }
});
const description$1 = "_description_f14k1_123";
const editInputButton$1 = "_editInputButton_f14k1_129";
const descriptionText = "_descriptionText_f14k1_133";
const icon$1 = "_icon_f14k1_139";
const style0$9 = {
  description: description$1,
  editInputButton: editInputButton$1,
  descriptionText,
  icon: icon$1
};
const cssModules$9 = {
  "$style": style0$9
};
const DescriptionInput = /* @__PURE__ */ _export_sfc(_sfc_main$a, [["__cssModules", cssModules$9]]);
const _sfc_main$9 = /* @__PURE__ */ defineComponent({
  __name: "HeaderSection",
  props: /* @__PURE__ */ mergeModels({
    hasRuns: { type: Boolean },
    isSaving: { type: Boolean },
    showConfig: { type: Boolean },
    runTestEnabled: { type: Boolean },
    startEditing: { type: Function },
    saveChanges: { type: Function },
    handleKeydown: { type: Function },
    onSaveTest: { type: Function },
    runTest: { type: Function },
    toggleConfig: { type: Function },
    getFieldIssues: { type: Function }
  }, {
    "name": { required: true },
    "nameModifiers": {},
    "description": { required: true },
    "descriptionModifiers": {}
  }),
  emits: ["update:name", "update:description"],
  setup(__props) {
    const props = __props;
    const name2 = useModel(__props, "name");
    const description2 = useModel(__props, "description");
    const locale = useI18n();
    const showSavingIndicator = computed(() => {
      return !name2.value.isEditing;
    });
    return (_ctx, _cache) => {
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_N8nButton = resolveComponent("N8nButton");
      const _component_N8nTooltip = resolveComponent("N8nTooltip");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.headerSection)
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.headerMeta)
        }, [
          createBaseVNode("div", {
            class: normalizeClass(_ctx.$style.name)
          }, [
            createVNode(_component_n8n_icon_button, {
              class: normalizeClass(_ctx.$style.backButton),
              icon: "arrow-left",
              type: "tertiary",
              title: unref(locale).baseText("testDefinition.edit.backButtonTitle"),
              onClick: _cache[0] || (_cache[0] = ($event) => _ctx.$router.back())
            }, null, 8, ["class", "title"]),
            createVNode(TestNameInput, {
              modelValue: name2.value,
              "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => name2.value = $event),
              class: normalizeClass({ "has-issues": _ctx.getFieldIssues("name").length > 0 }),
              "start-editing": _ctx.startEditing,
              "save-changes": _ctx.saveChanges,
              "handle-keydown": _ctx.handleKeydown
            }, null, 8, ["modelValue", "class", "start-editing", "save-changes", "handle-keydown"]),
            showSavingIndicator.value ? (openBlock(), createElementBlock("div", {
              key: 0,
              class: normalizeClass(_ctx.$style.lastSaved)
            }, [
              _ctx.isSaving ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
                createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.saving")), 1)
              ], 64)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
                createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.saved")), 1)
              ], 64))
            ], 2)) : createCommentVNode("", true)
          ], 2),
          createVNode(DescriptionInput, {
            modelValue: description2.value,
            "onUpdate:modelValue": _cache[2] || (_cache[2] = ($event) => description2.value = $event),
            "start-editing": _ctx.startEditing,
            "save-changes": _ctx.saveChanges,
            "handle-keydown": _ctx.handleKeydown,
            class: normalizeClass(_ctx.$style.descriptionInput)
          }, null, 8, ["modelValue", "start-editing", "save-changes", "handle-keydown", "class"])
        ], 2),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.controls)
        }, [
          props.hasRuns ? (openBlock(), createBlock(_component_N8nButton, {
            key: 0,
            size: "small",
            icon: _ctx.showConfig ? "eye-slash" : "eye",
            "data-test-id": "toggle-config-button",
            label: _ctx.showConfig ? unref(locale).baseText("testDefinition.edit.hideConfig") : unref(locale).baseText("testDefinition.edit.showConfig"),
            type: "tertiary",
            onClick: _ctx.toggleConfig
          }, null, 8, ["icon", "label", "onClick"])) : createCommentVNode("", true),
          createVNode(_component_N8nTooltip, {
            disabled: _ctx.runTestEnabled,
            placement: "left"
          }, {
            content: withCtx(() => [
              renderSlot(_ctx.$slots, "runTestTooltip")
            ]),
            default: withCtx(() => [
              createVNode(_component_N8nButton, {
                disabled: !_ctx.runTestEnabled,
                class: normalizeClass(_ctx.$style.runTestButton),
                size: "small",
                "data-test-id": "run-test-button",
                label: unref(locale).baseText("testDefinition.runTest"),
                type: "primary",
                onClick: _ctx.runTest
              }, null, 8, ["disabled", "class", "label", "onClick"])
            ]),
            _: 3
          }, 8, ["disabled"])
        ], 2)
      ], 2);
    };
  }
});
const headerSection = "_headerSection_28wt7_123";
const headerMeta = "_headerMeta_28wt7_131";
const name = "_name_28wt7_135";
const lastSaved = "_lastSaved_28wt7_140";
const descriptionInput = "_descriptionInput_28wt7_145";
const controls = "_controls_28wt7_149";
const backButton = "_backButton_28wt7_154";
const style0$8 = {
  headerSection,
  headerMeta,
  name,
  lastSaved,
  descriptionInput,
  controls,
  backButton
};
const cssModules$8 = {
  "$style": style0$8
};
const HeaderSection = /* @__PURE__ */ _export_sfc(_sfc_main$9, [["__cssModules", cssModules$8]]);
const _sfc_main$8 = /* @__PURE__ */ defineComponent({
  __name: "RunsSection",
  props: /* @__PURE__ */ mergeModels({
    runs: {},
    testId: {},
    appliedTheme: {}
  }, {
    "selectedMetric": { required: true },
    "selectedMetricModifiers": {}
  }),
  emits: /* @__PURE__ */ mergeModels(["deleteRuns"], ["update:selectedMetric"]),
  setup(__props, { emit: __emit }) {
    const emit = __emit;
    const selectedMetric = useModel(__props, "selectedMetric");
    function onDeleteRuns(toDelete) {
      emit("deleteRuns", toDelete);
    }
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.runs)
      }, [
        createVNode(MetricsChart, {
          selectedMetric: selectedMetric.value,
          "onUpdate:selectedMetric": _cache[0] || (_cache[0] = ($event) => selectedMetric.value = $event),
          runs: _ctx.runs,
          theme: _ctx.appliedTheme
        }, null, 8, ["selectedMetric", "runs", "theme"]),
        createVNode(TestRunsTable, {
          class: normalizeClass(_ctx.$style.runsTable),
          runs: _ctx.runs,
          selectable: true,
          "data-test-id": "past-runs-table",
          onDeleteRuns
        }, null, 8, ["class", "runs"])
      ], 2);
    };
  }
});
const runs = "_runs_1llgd_123";
const style0$7 = {
  runs
};
const cssModules$7 = {
  "$style": style0$7
};
const RunsSection = /* @__PURE__ */ _export_sfc(_sfc_main$8, [["__cssModules", cssModules$7]]);
const _sfc_main$7 = /* @__PURE__ */ defineComponent({
  __name: "NodesPinning",
  props: {
    modelValue: {}
  },
  emits: ["update:modelValue"],
  setup(__props, { emit: __emit }) {
    const workflowsStore = useWorkflowsStore();
    const nodeTypesStore = useNodeTypesStore();
    const route = useRoute();
    const router = useRouter();
    const locale = useI18n();
    const { resetWorkspace, initializeWorkspace } = useCanvasOperations({ router });
    const eventBus = createEventBus();
    const style = useCssModule();
    const uuid = crypto.randomUUID();
    const props = __props;
    const emit = __emit;
    const isLoading = ref(true);
    const workflowId = computed(() => route.params.name);
    const testId = computed(() => route.params.testId);
    const workflow = computed(() => workflowsStore.getWorkflowById(workflowId.value));
    const workflowObject = computed(() => workflowsStore.getCurrentWorkflow(true));
    const canvasId = computed(() => `${uuid}-${testId.value}`);
    const { onNodesInitialized, fitView, zoomTo } = useVueFlow({ id: canvasId.value });
    const nodes = computed(() => {
      return workflow.value.nodes ?? [];
    });
    const connections = computed(() => workflow.value.connections);
    const { nodes: mappedNodes, connections: mappedConnections } = useCanvasMapping({
      nodes,
      connections,
      workflowObject
    });
    async function loadData() {
      workflowsStore.resetState();
      resetWorkspace();
      const loadingPromise = Promise.all([
        nodeTypesStore.getNodeTypes(),
        workflowsStore.fetchWorkflow(workflowId.value)
      ]);
      await loadingPromise;
      initializeWorkspace(workflow.value);
      disableAllNodes();
    }
    function getNodeNameById(id) {
      return mappedNodes.value.find((node) => node.id === id)?.data?.name;
    }
    function updateNodeClasses(nodeIds, isPinned) {
      eventBus.emit("nodes:action", {
        ids: nodeIds,
        action: "update:node:class",
        payload: {
          className: style.pinnedNode,
          add: isPinned
        }
      });
      eventBus.emit("nodes:action", {
        ids: nodeIds,
        action: "update:node:class",
        payload: {
          className: style.notPinnedNode,
          add: !isPinned
        }
      });
    }
    function disableAllNodes() {
      const ids = mappedNodes.value.map((node) => node.id);
      updateNodeClasses(ids, false);
      const pinnedNodes = props.modelValue.map((node) => node.id).filter((id) => id !== null);
      if (pinnedNodes.length > 0) {
        updateNodeClasses(pinnedNodes, true);
      }
    }
    function onPinButtonClick(data) {
      const nodeName = getNodeNameById(data.id);
      if (!nodeName) return;
      const isPinned = props.modelValue.some((node) => node.id === data.id);
      const updatedNodes = isPinned ? props.modelValue.filter((node) => node.id !== data.id) : [...props.modelValue, { name: nodeName, id: data.id }];
      emit("update:modelValue", updatedNodes);
      updateNodeClasses([data.id], !isPinned);
    }
    function isPinButtonVisible(outputs) {
      return outputs.length === 1;
    }
    onNodesInitialized(async () => {
      await fitView();
      isLoading.value = false;
      await zoomTo(0.7, { duration: 400 });
    });
    onMounted(loadData);
    return (_ctx, _cache) => {
      const _component_N8nHeading = resolveComponent("N8nHeading");
      const _component_N8nText = resolveComponent("N8nText");
      const _component_N8nSpinner = resolveComponent("N8nSpinner");
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_Canvas = __unplugin_components_0$1;
      return unref(mappedNodes).length === 0 ? (openBlock(), createElementBlock("div", {
        key: 0,
        class: normalizeClass(_ctx.$style.noNodes)
      }, [
        createVNode(_component_N8nHeading, {
          size: "large",
          bold: true,
          class: normalizeClass(_ctx.$style.noNodesTitle)
        }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.pinNodes.noNodes.title")), 1)
          ]),
          _: 1
        }, 8, ["class"]),
        createVNode(_component_N8nText, null, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.pinNodes.noNodes.description")), 1)
          ]),
          _: 1
        })
      ], 2)) : (openBlock(), createElementBlock("div", {
        key: 1,
        class: normalizeClass(_ctx.$style.container)
      }, [
        isLoading.value ? (openBlock(), createBlock(_component_N8nSpinner, {
          key: 0,
          size: "xlarge",
          type: "dots",
          class: normalizeClass(_ctx.$style.spinner)
        }, null, 8, ["class"])) : createCommentVNode("", true),
        createVNode(_component_Canvas, {
          id: canvasId.value,
          loading: isLoading.value,
          class: normalizeClass({ [_ctx.$style.canvas]: true }),
          nodes: unref(mappedNodes),
          connections: unref(mappedConnections),
          "show-bug-reporting-button": false,
          "read-only": true,
          "event-bus": unref(eventBus)
        }, {
          nodeToolbar: withCtx(({ data, outputs }) => [
            createBaseVNode("div", {
              class: normalizeClass(_ctx.$style.pinButtonContainer)
            }, [
              isPinButtonVisible(outputs) ? (openBlock(), createBlock(unref(N8nTooltip), {
                key: 0,
                placement: "left"
              }, {
                content: withCtx(() => [
                  createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.nodesPinning.pinButtonTooltip")), 1)
                ]),
                default: withCtx(() => [
                  createVNode(_component_n8n_icon_button, {
                    type: "tertiary",
                    size: "large",
                    icon: "thumbtack",
                    class: normalizeClass(_ctx.$style.pinButton),
                    "data-test-id": "node-pin-button",
                    onClick: ($event) => onPinButtonClick(data)
                  }, null, 8, ["class", "onClick"])
                ]),
                _: 2
              }, 1024)) : createCommentVNode("", true)
            ], 2)
          ]),
          _: 1
        }, 8, ["id", "loading", "class", "nodes", "connections", "event-bus"])
      ], 2));
    };
  }
});
const container$1 = "_container_108jn_123";
const pinButtonContainer = "_pinButtonContainer_108jn_128";
const pinButton = "_pinButton_108jn_128";
const notPinnedNode = "_notPinnedNode_108jn_142";
const pinnedNode = "_pinnedNode_108jn_143";
const spinner = "_spinner_108jn_154";
const noNodes = "_noNodes_108jn_161";
const style0$6 = {
  container: container$1,
  pinButtonContainer,
  pinButton,
  notPinnedNode,
  pinnedNode,
  spinner,
  noNodes
};
const cssModules$6 = {
  "$style": style0$6
};
const __unplugin_components_1 = /* @__PURE__ */ _export_sfc(_sfc_main$7, [["__cssModules", cssModules$6]]);
const arrowConnector = "_arrowConnector_1mi41_123";
const style0$5 = {
  arrowConnector
};
const _sfc_main$6 = {};
function _sfc_render(_ctx, _cache) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.$style.arrowConnector)
  }, null, 2);
}
const cssModules$5 = {
  "$style": style0$5
};
const __unplugin_components_0 = /* @__PURE__ */ _export_sfc(_sfc_main$6, [["render", _sfc_render], ["__cssModules", cssModules$5]]);
const _hoisted_1$1 = ["aria-expanded", "aria-controls"];
const _sfc_main$5 = /* @__PURE__ */ defineComponent({
  __name: "EvaluationStep",
  props: {
    title: {},
    warning: { type: Boolean, default: false },
    small: { type: Boolean, default: false },
    expanded: { type: Boolean, default: true },
    description: { default: "" },
    issues: { default: () => [] },
    showIssues: { type: Boolean, default: true }
  },
  setup(__props) {
    const props = __props;
    const locale = useI18n();
    const isExpanded = ref(props.expanded);
    const contentRef = ref(null);
    const containerRef = ref(null);
    const toggleExpand = async () => {
      isExpanded.value = !isExpanded.value;
      if (isExpanded.value) {
        await nextTick();
        if (containerRef.value) {
          containerRef.value.style.height = "auto";
        }
      }
    };
    return (_ctx, _cache) => {
      const _component_N8nInfoTip = resolveComponent("N8nInfoTip");
      const _component_font_awesome_icon = resolveComponent("font-awesome-icon");
      return openBlock(), createElementBlock("div", {
        ref_key: "containerRef",
        ref: containerRef,
        class: normalizeClass([_ctx.$style.evaluationStep, _ctx.small && _ctx.$style.small]),
        "data-test-id": "evaluation-step"
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.content)
        }, [
          createBaseVNode("div", {
            class: normalizeClass(_ctx.$style.header)
          }, [
            createBaseVNode("div", {
              class: normalizeClass([_ctx.$style.icon, _ctx.warning && _ctx.$style.warning])
            }, [
              renderSlot(_ctx.$slots, "icon")
            ], 2),
            createBaseVNode("h3", {
              class: normalizeClass(_ctx.$style.title)
            }, toDisplayString(_ctx.title), 3),
            _ctx.issues.length > 0 && _ctx.showIssues ? (openBlock(), createElementBlock("span", {
              key: 0,
              class: normalizeClass(_ctx.$style.warningIcon)
            }, [
              createVNode(_component_N8nInfoTip, {
                bold: true,
                type: "tooltip",
                theme: "warning",
                "tooltip-placement": "right"
              }, {
                default: withCtx(() => [
                  createTextVNode(toDisplayString(_ctx.issues.map((issue) => issue.message).join(", ")), 1)
                ]),
                _: 1
              })
            ], 2)) : createCommentVNode("", true),
            _ctx.$slots.cardContent ? (openBlock(), createElementBlock("button", {
              key: 1,
              class: normalizeClass(_ctx.$style.collapseButton),
              "aria-expanded": isExpanded.value,
              "aria-controls": "content-" + _ctx.title.replace(/\s+/g, "-"),
              "data-test-id": "evaluation-step-collapse-button",
              onClick: toggleExpand
            }, [
              createTextVNode(toDisplayString(isExpanded.value ? unref(locale).baseText("testDefinition.edit.step.collapse") : unref(locale).baseText("testDefinition.edit.step.expand")) + " ", 1),
              createVNode(_component_font_awesome_icon, {
                icon: isExpanded.value ? "angle-down" : "angle-right",
                size: "lg"
              }, null, 8, ["icon"])
            ], 10, _hoisted_1$1)) : createCommentVNode("", true)
          ], 2),
          _ctx.description ? (openBlock(), createElementBlock("div", {
            key: 0,
            class: normalizeClass(_ctx.$style.description)
          }, toDisplayString(_ctx.description), 3)) : createCommentVNode("", true),
          _ctx.$slots.cardContent ? (openBlock(), createBlock(unref(ElCollapseTransition), { key: 1 }, {
            default: withCtx(() => [
              withDirectives(createBaseVNode("div", {
                class: normalizeClass(_ctx.$style.cardContentWrapper)
              }, [
                createBaseVNode("div", {
                  ref_key: "contentRef",
                  ref: contentRef,
                  class: normalizeClass(_ctx.$style.cardContent),
                  "data-test-id": "evaluation-step-content"
                }, [
                  renderSlot(_ctx.$slots, "cardContent")
                ], 2)
              ], 2), [
                [vShow, isExpanded.value]
              ])
            ]),
            _: 3
          })) : createCommentVNode("", true)
        ], 2)
      ], 2);
    };
  }
});
const evaluationStep = "_evaluationStep_65xdl_123";
const small = "_small_65xdl_135";
const icon = "_icon_65xdl_140";
const warning = "_warning_65xdl_149";
const content$1 = "_content_65xdl_153";
const header = "_header_65xdl_158";
const title = "_title_65xdl_164";
const warningIcon = "_warningIcon_65xdl_170";
const cardContent = "_cardContent_65xdl_174";
const collapseButton = "_collapseButton_65xdl_179";
const cardContentWrapper = "_cardContentWrapper_65xdl_192";
const description = "_description_65xdl_196";
const style0$4 = {
  evaluationStep,
  small,
  icon,
  warning,
  content: content$1,
  header,
  title,
  warningIcon,
  cardContent,
  collapseButton,
  cardContentWrapper,
  description
};
const cssModules$4 = {
  "$style": style0$4
};
const EvaluationStep = /* @__PURE__ */ _export_sfc(_sfc_main$5, [["__cssModules", cssModules$4]]);
const _hoisted_1 = { "data-test-id": "workflow-tags-field" };
const _sfc_main$4 = /* @__PURE__ */ defineComponent({
  __name: "TagsInput",
  props: {
    modelValue: { default: () => ({
      isEditing: false,
      value: [],
      tempValue: []
    }) },
    allTags: {},
    tagsById: {},
    isLoading: { type: Boolean },
    startEditing: {},
    saveChanges: {},
    cancelEditing: {},
    createTag: { type: Function, default: void 0 }
  },
  emits: ["update:modelValue"],
  setup(__props, { emit: __emit }) {
    const props = __props;
    const emit = __emit;
    const locale = useI18n();
    const tagsEventBus = createEventBus();
    const getTagName = computed(() => (tagId) => {
      return props.tagsById[tagId]?.name ?? "";
    });
    function updateTags(tags) {
      emit("update:modelValue", {
        ...props.modelValue,
        tempValue: tags
      });
    }
    return (_ctx, _cache) => {
      const _component_n8n_text = resolveComponent("n8n-text");
      const _component_n8n_tag = resolveComponent("n8n-tag");
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_TagsDropdown = _sfc_main$c;
      const _component_n8n_input_label = resolveComponent("n8n-input-label");
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createVNode(_component_n8n_input_label, {
          label: unref(locale).baseText("testDefinition.edit.tagName"),
          bold: false,
          size: "small"
        }, {
          default: withCtx(() => [
            !_ctx.modelValue.isEditing ? (openBlock(), createElementBlock("div", {
              key: 0,
              class: normalizeClass(_ctx.$style.tagsRead),
              onClick: _cache[0] || (_cache[0] = ($event) => _ctx.startEditing("tags"))
            }, [
              _ctx.modelValue.value.length === 0 ? (openBlock(), createBlock(_component_n8n_text, {
                key: 0,
                size: "small"
              }, {
                default: withCtx(() => [
                  createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.selectTag")), 1)
                ]),
                _: 1
              })) : createCommentVNode("", true),
              (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.modelValue.value, (tagId) => {
                return openBlock(), createBlock(_component_n8n_tag, {
                  key: tagId,
                  text: getTagName.value(tagId),
                  "data-test-id": "evaluation-tag-field"
                }, null, 8, ["text"]);
              }), 128)),
              createVNode(_component_n8n_icon_button, {
                class: normalizeClass(_ctx.$style.editInputButton),
                icon: "pen",
                type: "tertiary",
                size: "small",
                transparent: ""
              }, null, 8, ["class"])
            ], 2)) : (openBlock(), createBlock(_component_TagsDropdown, {
              key: 1,
              "model-value": _ctx.modelValue.tempValue,
              placeholder: unref(locale).baseText("executionAnnotationView.chooseOrCreateATag"),
              "create-enabled": _ctx.modelValue.tempValue.length === 0,
              "all-tags": _ctx.allTags,
              "is-loading": _ctx.isLoading,
              "tags-by-id": _ctx.tagsById,
              "data-test-id": "workflow-tags-dropdown",
              "event-bus": unref(tagsEventBus),
              "create-tag": _ctx.createTag,
              "manage-enabled": false,
              "multiple-limit": 1,
              "onUpdate:modelValue": updateTags,
              onEsc: _cache[1] || (_cache[1] = ($event) => _ctx.cancelEditing("tags")),
              onBlur: _cache[2] || (_cache[2] = ($event) => _ctx.saveChanges("tags"))
            }, null, 8, ["model-value", "placeholder", "create-enabled", "all-tags", "is-loading", "tags-by-id", "event-bus", "create-tag"]))
          ]),
          _: 1
        }, 8, ["label"])
      ]);
    };
  }
});
const tagsRead = "_tagsRead_z5pm4_123";
const editInputButton = "_editInputButton_z5pm4_123";
const style0$3 = {
  tagsRead,
  editInputButton
};
const cssModules$3 = {
  "$style": style0$3
};
const TagsInput = /* @__PURE__ */ _export_sfc(_sfc_main$4, [["__cssModules", cssModules$3]]);
const _sfc_main$3 = /* @__PURE__ */ defineComponent({
  __name: "WorkflowSelector",
  props: {
    modelValue: { default: () => ({
      mode: "id",
      value: "",
      __rl: true
    }) },
    examplePinnedData: { default: () => ({}) },
    sampleWorkflowName: { default: void 0 }
  },
  emits: ["update:modelValue"],
  setup(__props) {
    const props = __props;
    const locale = useI18n();
    const subworkflowName = computed(() => {
      if (props.sampleWorkflowName) {
        return locale.baseText("testDefinition.workflowInput.subworkflowName", {
          interpolate: { name: props.sampleWorkflowName }
        });
      }
      return locale.baseText("testDefinition.workflowInput.subworkflowName.default");
    });
    const sampleWorkflow = computed(() => {
      return {
        ...SAMPLE_EVALUATION_WORKFLOW,
        name: subworkflowName.value,
        pinData: props.examplePinnedData
      };
    });
    return (_ctx, _cache) => {
      const _component_WorkflowSelectorParameterInput = __unplugin_components_0$2;
      const _component_n8n_input_label = resolveComponent("n8n-input-label");
      return openBlock(), createElementBlock("div", null, [
        createVNode(_component_n8n_input_label, {
          label: unref(locale).baseText("testDefinition.edit.workflowSelectorLabel"),
          bold: false
        }, {
          default: withCtx(() => [
            createVNode(_component_WorkflowSelectorParameterInput, {
              ref: "workflowInput",
              parameter: {
                displayName: unref(locale).baseText("testDefinition.edit.workflowSelectorDisplayName"),
                name: "workflowId",
                type: "workflowSelector",
                default: ""
              },
              "model-value": _ctx.modelValue,
              "display-title": unref(locale).baseText("testDefinition.edit.workflowSelectorTitle"),
              "is-value-expression": false,
              "expression-edit-dialog-visible": false,
              path: "workflows",
              "allow-new": "",
              "sample-workflow": sampleWorkflow.value,
              "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => _ctx.$emit("update:modelValue", $event))
            }, null, 8, ["parameter", "model-value", "display-title", "sample-workflow"])
          ]),
          _: 1
        }, 8, ["label"])
      ]);
    };
  }
});
const _sfc_main$2 = /* @__PURE__ */ defineComponent({
  __name: "MetricsInput",
  props: {
    modelValue: {}
  },
  emits: ["update:modelValue", "deleteMetric"],
  setup(__props, { emit: __emit }) {
    const props = __props;
    const emit = __emit;
    const locale = useI18n();
    function addNewMetric() {
      emit("update:modelValue", [...props.modelValue, { name: "" }]);
    }
    function updateMetric(index, name2) {
      const newMetrics = [...props.modelValue];
      newMetrics[index].name = name2;
      emit("update:modelValue", newMetrics);
    }
    function onDeleteMetric(metric) {
      emit("deleteMetric", metric);
    }
    return (_ctx, _cache) => {
      const _component_N8nInput = resolveComponent("N8nInput");
      const _component_n8n_icon_button = resolveComponent("n8n-icon-button");
      const _component_n8n_button = resolveComponent("n8n-button");
      const _component_n8n_input_label = resolveComponent("n8n-input-label");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass([_ctx.$style.metrics])
      }, [
        createVNode(_component_n8n_input_label, {
          label: unref(locale).baseText("testDefinition.edit.metricsFields"),
          bold: false,
          class: normalizeClass(_ctx.$style.metricField)
        }, {
          default: withCtx(() => [
            createBaseVNode("div", {
              class: normalizeClass(_ctx.$style.metricsContainer)
            }, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.modelValue, (metric, index) => {
                return openBlock(), createElementBlock("div", {
                  key: index,
                  class: normalizeClass(_ctx.$style.metricItem)
                }, [
                  createVNode(_component_N8nInput, {
                    ref_for: true,
                    ref: `metric_${index}`,
                    "data-test-id": "evaluation-metric-item",
                    "model-value": metric.name,
                    placeholder: unref(locale).baseText("testDefinition.edit.metricsPlaceholder"),
                    "onUpdate:modelValue": (value) => updateMetric(index, value)
                  }, null, 8, ["model-value", "placeholder", "onUpdate:modelValue"]),
                  createVNode(_component_n8n_icon_button, {
                    icon: "trash",
                    type: "text",
                    onClick: ($event) => onDeleteMetric(metric)
                  }, null, 8, ["onClick"])
                ], 2);
              }), 128)),
              createVNode(_component_n8n_button, {
                type: "tertiary",
                label: unref(locale).baseText("testDefinition.edit.metricsNew"),
                class: normalizeClass(_ctx.$style.newMetricButton),
                onClick: addNewMetric
              }, null, 8, ["label", "class"])
            ], 2)
          ]),
          _: 1
        }, 8, ["label", "class"])
      ], 2);
    };
  }
});
const metricsContainer = "_metricsContainer_1eaf8_123";
const metricItem = "_metricItem_1eaf8_129";
const metricField = "_metricField_1eaf8_134";
const metricsDivider = "_metricsDivider_1eaf8_139";
const newMetricButton = "_newMetricButton_1eaf8_144";
const style0$2 = {
  metricsContainer,
  metricItem,
  metricField,
  metricsDivider,
  newMetricButton
};
const cssModules$2 = {
  "$style": style0$2
};
const MetricsInput = /* @__PURE__ */ _export_sfc(_sfc_main$2, [["__cssModules", cssModules$2]]);
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  __name: "ConfigSection",
  props: /* @__PURE__ */ mergeModels({
    showConfig: { type: Boolean },
    tagUsageCount: {},
    allTags: {},
    tagsById: {},
    isLoading: { type: Boolean },
    examplePinnedData: {},
    sampleWorkflowName: {},
    getFieldIssues: { type: Function },
    startEditing: { type: Function },
    saveChanges: { type: Function },
    cancelEditing: { type: Function },
    createTag: { type: Function }
  }, {
    "tags": { required: true },
    "tagsModifiers": {},
    "evaluationWorkflow": { required: true },
    "evaluationWorkflowModifiers": {},
    "metrics": { required: true },
    "metricsModifiers": {},
    "mockedNodes": {
      required: true
    },
    "mockedNodesModifiers": {}
  }),
  emits: /* @__PURE__ */ mergeModels(["openPinningModal", "deleteMetric"], ["update:tags", "update:evaluationWorkflow", "update:metrics", "update:mockedNodes"]),
  setup(__props, { emit: __emit }) {
    const emit = __emit;
    const locale = useI18n();
    const changedFieldsKeys = ref([]);
    const tags = useModel(__props, "tags");
    const evaluationWorkflow = useModel(
      __props,
      "evaluationWorkflow"
    );
    const metrics = useModel(__props, "metrics");
    const mockedNodes = useModel(__props, "mockedNodes");
    const nodePinningModal = ref(null);
    function updateChangedFieldsKeys(key) {
      changedFieldsKeys.value.push(key);
    }
    function showFieldIssues(fieldKey) {
      return changedFieldsKeys.value.includes(fieldKey);
    }
    return (_ctx, _cache) => {
      const _component_BlockArrow = __unplugin_components_0;
      const _component_font_awesome_icon = resolveComponent("font-awesome-icon");
      const _component_n8n_button = resolveComponent("n8n-button");
      const _component_N8nHeading = resolveComponent("N8nHeading");
      const _component_NodesPinning = __unplugin_components_1;
      const _component_Modal = __unplugin_components_2;
      return openBlock(), createElementBlock("div", {
        class: normalizeClass([_ctx.$style.panelBlock, { [_ctx.$style.hidden]: !_ctx.showConfig }])
      }, [
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.panelIntro)
        }, toDisplayString(unref(locale).baseText("testDefinition.edit.step.intro")), 3),
        createVNode(_component_BlockArrow, {
          class: normalizeClass(_ctx.$style.introArrow)
        }, null, 8, ["class"]),
        createVNode(EvaluationStep, {
          class: normalizeClass(_ctx.$style.step),
          title: unref(locale).baseText("testDefinition.edit.step.executions", {
            adjustToNumber: _ctx.tagUsageCount
          }),
          description: unref(locale).baseText("testDefinition.edit.step.executions.description"),
          issues: _ctx.getFieldIssues("tags"),
          "show-issues": showFieldIssues("tags")
        }, {
          icon: withCtx(() => [
            createVNode(_component_font_awesome_icon, {
              icon: "history",
              size: "lg"
            })
          ]),
          cardContent: withCtx(() => [
            createVNode(TagsInput, {
              modelValue: tags.value,
              "onUpdate:modelValue": [
                _cache[0] || (_cache[0] = ($event) => tags.value = $event),
                _cache[1] || (_cache[1] = ($event) => updateChangedFieldsKeys("tags"))
              ],
              class: normalizeClass({ "has-issues": _ctx.getFieldIssues("tags") }),
              "all-tags": _ctx.allTags,
              "tags-by-id": _ctx.tagsById,
              "is-loading": _ctx.isLoading,
              "start-editing": _ctx.startEditing,
              "save-changes": _ctx.saveChanges,
              "cancel-editing": _ctx.cancelEditing,
              "create-tag": _ctx.createTag
            }, null, 8, ["modelValue", "class", "all-tags", "tags-by-id", "is-loading", "start-editing", "save-changes", "cancel-editing", "create-tag"])
          ]),
          _: 1
        }, 8, ["class", "title", "description", "issues", "show-issues"]),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.evaluationArrows)
        }, [
          createVNode(_component_BlockArrow),
          createVNode(_component_BlockArrow)
        ], 2),
        createVNode(EvaluationStep, {
          class: normalizeClass(_ctx.$style.step),
          title: unref(locale).baseText("testDefinition.edit.step.mockedNodes", {
            adjustToNumber: mockedNodes.value?.length ?? 0
          }),
          small: true,
          expanded: true,
          description: unref(locale).baseText("testDefinition.edit.step.nodes.description"),
          issues: _ctx.getFieldIssues("mockedNodes"),
          "show-issues": showFieldIssues("mockedNodes")
        }, {
          icon: withCtx(() => [
            createVNode(_component_font_awesome_icon, {
              icon: "thumbtack",
              size: "lg"
            })
          ]),
          cardContent: withCtx(() => [
            createVNode(_component_n8n_button, {
              size: "small",
              "data-test-id": "select-nodes-button",
              label: unref(locale).baseText("testDefinition.edit.selectNodes"),
              type: "tertiary",
              onClick: _cache[2] || (_cache[2] = ($event) => _ctx.$emit("openPinningModal"))
            }, null, 8, ["label"])
          ]),
          _: 1
        }, 8, ["class", "title", "description", "issues", "show-issues"]),
        createVNode(EvaluationStep, {
          class: normalizeClass(_ctx.$style.step),
          title: unref(locale).baseText("testDefinition.edit.step.reRunExecutions"),
          small: true,
          description: unref(locale).baseText("testDefinition.edit.step.reRunExecutions.description")
        }, {
          icon: withCtx(() => [
            createVNode(_component_font_awesome_icon, {
              icon: "redo",
              size: "lg"
            })
          ]),
          _: 1
        }, 8, ["class", "title", "description"]),
        createVNode(EvaluationStep, {
          class: normalizeClass(_ctx.$style.step),
          title: unref(locale).baseText("testDefinition.edit.step.compareExecutions"),
          description: unref(locale).baseText("testDefinition.edit.step.compareExecutions.description"),
          issues: _ctx.getFieldIssues("evaluationWorkflow"),
          "show-issues": showFieldIssues("evaluationWorkflow")
        }, {
          icon: withCtx(() => [
            createVNode(_component_font_awesome_icon, {
              icon: "equals",
              size: "lg"
            })
          ]),
          cardContent: withCtx(() => [
            createVNode(_sfc_main$3, {
              modelValue: evaluationWorkflow.value,
              "onUpdate:modelValue": [
                _cache[3] || (_cache[3] = ($event) => evaluationWorkflow.value = $event),
                _cache[4] || (_cache[4] = ($event) => updateChangedFieldsKeys("evaluationWorkflow"))
              ],
              "example-pinned-data": _ctx.examplePinnedData,
              class: normalizeClass({ "has-issues": _ctx.getFieldIssues("evaluationWorkflow").length > 0 }),
              "sample-workflow-name": _ctx.sampleWorkflowName
            }, null, 8, ["modelValue", "example-pinned-data", "class", "sample-workflow-name"])
          ]),
          _: 1
        }, 8, ["class", "title", "description", "issues", "show-issues"]),
        createVNode(EvaluationStep, {
          class: normalizeClass(_ctx.$style.step),
          title: unref(locale).baseText("testDefinition.edit.step.metrics"),
          description: unref(locale).baseText("testDefinition.edit.step.metrics.description"),
          issues: _ctx.getFieldIssues("metrics"),
          "show-issues": showFieldIssues("metrics")
        }, {
          icon: withCtx(() => [
            createVNode(_component_font_awesome_icon, {
              icon: "chart-bar",
              size: "lg"
            })
          ]),
          cardContent: withCtx(() => [
            createVNode(MetricsInput, {
              modelValue: metrics.value,
              "onUpdate:modelValue": [
                _cache[5] || (_cache[5] = ($event) => metrics.value = $event),
                _cache[7] || (_cache[7] = ($event) => updateChangedFieldsKeys("metrics"))
              ],
              class: normalizeClass({ "has-issues": _ctx.getFieldIssues("metrics").length > 0 }),
              onDeleteMetric: _cache[6] || (_cache[6] = (metric) => emit("deleteMetric", metric))
            }, null, 8, ["modelValue", "class"])
          ]),
          _: 1
        }, 8, ["class", "title", "description", "issues", "show-issues"]),
        createVNode(_component_Modal, {
          ref_key: "nodePinningModal",
          ref: nodePinningModal,
          width: "80vw",
          height: "85vh",
          name: unref(NODE_PINNING_MODAL_KEY)
        }, {
          header: withCtx(() => [
            createVNode(_component_N8nHeading, {
              size: "large",
              bold: true,
              class: normalizeClass(_ctx.$style.runsTableHeading)
            }, {
              default: withCtx(() => [
                createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.edit.selectNodes")), 1)
              ]),
              _: 1
            }, 8, ["class"])
          ]),
          content: withCtx(() => [
            createVNode(_component_NodesPinning, {
              modelValue: mockedNodes.value,
              "onUpdate:modelValue": _cache[8] || (_cache[8] = ($event) => mockedNodes.value = $event),
              "data-test-id": "nodes-pinning-modal"
            }, null, 8, ["modelValue"])
          ]),
          _: 1
        }, 8, ["name"])
      ], 2);
    };
  }
});
const panelBlock = "_panelBlock_1xhy8_123";
const hidden = "_hidden_1xhy8_133";
const noRuns$1 = "_noRuns_1xhy8_139";
const panelIntro = "_panelIntro_1xhy8_143";
const step = "_step_1xhy8_151";
const introArrow = "_introArrow_1xhy8_158";
const evaluationArrows = "_evaluationArrows_1xhy8_164";
const style0$1 = {
  panelBlock,
  hidden,
  noRuns: noRuns$1,
  panelIntro,
  step,
  introArrow,
  evaluationArrows
};
const cssModules$1 = {
  "$style": style0$1
};
const ConfigSection = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["__cssModules", cssModules$1]]);
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "TestDefinitionEditView",
  props: {
    testId: {}
  },
  setup(__props) {
    const props = __props;
    const router = useRouter();
    const route = useRoute();
    const locale = useI18n();
    const { debounce } = useDebounce();
    const toast = useToast();
    const testDefinitionStore = useTestDefinitionStore();
    const tagsStore = useAnnotationTagsStore();
    const uiStore = useUIStore();
    const executionsStore = useExecutionsStore();
    const workflowStore = useWorkflowsStore();
    const {
      state,
      isSaving,
      cancelEditing,
      loadTestData,
      createTest,
      updateTest,
      startEditing,
      saveChanges,
      handleKeydown,
      deleteMetric,
      updateMetrics
    } = useTestDefinitionForm();
    const isLoading = computed(() => tagsStore.isLoading);
    const allTags = computed(() => tagsStore.allTags);
    const tagsById = computed(() => tagsStore.tagsById);
    const testId = computed(() => props.testId ?? route.params.testId);
    const currentWorkflowId = computed(() => route.params.name);
    const appliedTheme = computed(() => uiStore.appliedTheme);
    const tagUsageCount = computed(
      () => tagsStore.tagsById[state.value.tags.value[0]]?.usageCount ?? 0
    );
    const workflowName = computed(() => workflowStore.workflow.name);
    const hasRuns = computed(() => runs2.value.length > 0);
    const fieldsIssues = computed(() => testDefinitionStore.getFieldIssues(testId.value) ?? []);
    const showConfig = ref(true);
    const selectedMetric = ref("");
    const examplePinnedData = ref({});
    onMounted(async () => {
      if (!testDefinitionStore.isFeatureEnabled) {
        toast.showMessage({
          title: locale.baseText("testDefinition.notImplemented"),
          type: "warning"
        });
        void router.push({
          name: VIEWS.WORKFLOW,
          params: { name: router.currentRoute.value.params.name }
        });
        return;
      }
      if (testId.value) {
        await loadTestData(testId.value);
      } else {
        await onSaveTest();
      }
    });
    async function onSaveTest() {
      try {
        let savedTest;
        if (testId.value) {
          savedTest = await updateTest(testId.value);
        } else {
          savedTest = await createTest(currentWorkflowId.value);
        }
        if (savedTest && route.name === VIEWS.NEW_TEST_DEFINITION) {
          await router.replace({
            name: VIEWS.TEST_DEFINITION_EDIT,
            params: { testId: savedTest.id }
          });
        }
      } catch (e) {
        toast.showError(e, locale.baseText("testDefinition.edit.testSaveFailed"));
      }
    }
    function getFieldIssues(key) {
      return fieldsIssues.value.filter((issue) => issue.field === key);
    }
    async function onDeleteMetric(deletedMetric) {
      if (deletedMetric.id) {
        await deleteMetric(deletedMetric.id, testId.value);
      }
    }
    async function handleCreateTag(tagName) {
      try {
        const newTag = await tagsStore.create(tagName);
        return newTag;
      } catch (error) {
        toast.showError(error, "Error", error.message);
        throw error;
      }
    }
    async function openPinningModal() {
      uiStore.openModal(NODE_PINNING_MODAL_KEY);
    }
    async function runTest() {
      await testDefinitionStore.startTestRun(testId.value);
      await testDefinitionStore.fetchTestRuns(testId.value);
    }
    const runs2 = computed(
      () => Object.values(testDefinitionStore.testRunsById ?? {}).filter(
        (run) => run.testDefinitionId === testId.value
      )
    );
    const isRunning = computed(() => runs2.value.some((run) => run.status === "running"));
    const isRunTestEnabled = computed(() => fieldsIssues.value.length === 0 && !isRunning.value);
    async function onDeleteRuns(toDelete) {
      await Promise.all(
        toDelete.map(async (run) => {
          await testDefinitionStore.deleteTestRun({ testDefinitionId: testId.value, runId: run.id });
        })
      );
    }
    function toggleConfig() {
      showConfig.value = !showConfig.value;
    }
    async function getExamplePinnedDataForTags() {
      const evaluationWorkflowExecutions = await executionsStore.fetchExecutions({
        workflowId: currentWorkflowId.value,
        annotationTags: state.value.tags.value
      });
      if (evaluationWorkflowExecutions.count > 0) {
        const firstExecution = evaluationWorkflowExecutions.results[0];
        const executionData = await executionsStore.fetchExecution(firstExecution.id);
        const resultData = executionData?.data?.resultData.runData;
        examplePinnedData.value = {
          "When called by a test run": [
            {
              json: {
                originalExecution: resultData,
                newExecution: resultData
              }
            }
          ]
        };
      }
    }
    watch(
      () => state.value.metrics,
      debounce(async () => await updateMetrics(testId.value), { debounceTime: 400 }),
      { deep: true }
    );
    watch(() => state.value.tags, getExamplePinnedDataForTags);
    watch(
      () => [
        state.value.description,
        state.value.name,
        state.value.tags,
        state.value.evaluationWorkflow,
        state.value.mockedNodes
      ],
      debounce(onSaveTest, { debounceTime: 400 }),
      { deep: true }
    );
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass([_ctx.$style.container, { [_ctx.$style.noRuns]: !hasRuns.value }])
      }, [
        createVNode(HeaderSection, {
          name: unref(state).name,
          "onUpdate:name": _cache[0] || (_cache[0] = ($event) => unref(state).name = $event),
          description: unref(state).description,
          "onUpdate:description": _cache[1] || (_cache[1] = ($event) => unref(state).description = $event),
          tags: unref(state).tags,
          "onUpdate:tags": _cache[2] || (_cache[2] = ($event) => unref(state).tags = $event),
          "has-runs": hasRuns.value,
          "is-saving": unref(isSaving),
          "get-field-issues": getFieldIssues,
          "start-editing": unref(startEditing),
          "save-changes": unref(saveChanges),
          "handle-keydown": unref(handleKeydown),
          "on-save-test": onSaveTest,
          "run-test": runTest,
          "show-config": showConfig.value,
          "toggle-config": toggleConfig,
          "run-test-enabled": isRunTestEnabled.value
        }, {
          runTestTooltip: withCtx(() => [
            fieldsIssues.value.length > 0 ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
              createBaseVNode("div", null, toDisplayString(unref(locale).baseText("testDefinition.completeConfig")), 1),
              (openBlock(true), createElementBlock(Fragment, null, renderList(fieldsIssues.value, (issue) => {
                return openBlock(), createElementBlock("div", {
                  key: issue.field
                }, "- " + toDisplayString(issue.message), 1);
              }), 128))
            ], 64)) : createCommentVNode("", true),
            isRunning.value ? (openBlock(), createElementBlock(Fragment, { key: 1 }, [
              createTextVNode(toDisplayString(unref(locale).baseText("testDefinition.testIsRunning")), 1)
            ], 64)) : createCommentVNode("", true)
          ]),
          _: 1
        }, 8, ["name", "description", "tags", "has-runs", "is-saving", "start-editing", "save-changes", "handle-keydown", "show-config", "run-test-enabled"]),
        createBaseVNode("div", {
          class: normalizeClass(_ctx.$style.content)
        }, [
          runs2.value.length > 0 ? (openBlock(), createBlock(RunsSection, {
            key: 0,
            selectedMetric: selectedMetric.value,
            "onUpdate:selectedMetric": _cache[3] || (_cache[3] = ($event) => selectedMetric.value = $event),
            runs: runs2.value,
            "test-id": testId.value,
            "applied-theme": appliedTheme.value,
            onDeleteRuns
          }, null, 8, ["selectedMetric", "runs", "test-id", "applied-theme"])) : createCommentVNode("", true),
          createVNode(ConfigSection, {
            tags: unref(state).tags,
            "onUpdate:tags": _cache[4] || (_cache[4] = ($event) => unref(state).tags = $event),
            evaluationWorkflow: unref(state).evaluationWorkflow,
            "onUpdate:evaluationWorkflow": _cache[5] || (_cache[5] = ($event) => unref(state).evaluationWorkflow = $event),
            metrics: unref(state).metrics,
            "onUpdate:metrics": _cache[6] || (_cache[6] = ($event) => unref(state).metrics = $event),
            mockedNodes: unref(state).mockedNodes,
            "onUpdate:mockedNodes": _cache[7] || (_cache[7] = ($event) => unref(state).mockedNodes = $event),
            "cancel-editing": unref(cancelEditing),
            "show-config": showConfig.value,
            "tag-usage-count": tagUsageCount.value,
            "all-tags": allTags.value,
            "tags-by-id": tagsById.value,
            "is-loading": isLoading.value,
            "get-field-issues": getFieldIssues,
            "start-editing": unref(startEditing),
            "save-changes": unref(saveChanges),
            "create-tag": handleCreateTag,
            "example-pinned-data": examplePinnedData.value,
            "sample-workflow-name": workflowName.value,
            onOpenPinningModal: openPinningModal,
            onDeleteMetric
          }, null, 8, ["tags", "evaluationWorkflow", "metrics", "mockedNodes", "cancel-editing", "show-config", "tag-usage-count", "all-tags", "tags-by-id", "is-loading", "start-editing", "save-changes", "example-pinned-data", "sample-workflow-name"])
        ], 2)
      ], 2);
    };
  }
});
const container = "_container_wdela_123";
const content = "_content_wdela_141";
const noRuns = "_noRuns_wdela_146";
const style0 = {
  container,
  content,
  noRuns
};
const cssModules = {
  "$style": style0
};
const TestDefinitionEditView = /* @__PURE__ */ _export_sfc(_sfc_main, [["__cssModules", cssModules]]);
export {
  TestDefinitionEditView as default
};
